likwid/doc/likwid-pin.1
.TH LIKWID-PIN 1 <DATE> likwid\-VERSION
.SH NAME
likwid-pin \- pin a sequential or threaded application to dedicated processors
.SH SYNOPSIS
.B likwid-pin 
.RB [\-vh]
.RB [ \-c
.IR corelist
.RB [ \-t
.IR OpenMP_type ]
.RB [ \-s
.IR skip_mask ]
.RB [ \-p]
.RB [ \-q]
.RB [ \-i]
.SH DESCRIPTION
.B likwid-pin
is a command line application to pin a sequential or multi threaded 
application to dedicated processors. It can be used as replacement for taskset. 
Opposite to taskset no affinity mask but single processors are specified.
For multi threaded applications based on the pthread library the 
.I pthread_create
library call is overloaded through a LD_PRELOAD and each created thread is pinned
to a dedicated processor as specified in 
.I core_list .
.PP
Per default every generated thread is pinned to the core in the order of calls 
to pthread_create. It is possible to skip single threads.
.PP
For OpenMP implementations gcc and icc compilers are supported.
Be aware that with pthreads the parent thread is always pinned. If you create for example 4 threads
with pthread_create and do not use the parent process as worker you still have to provide
num_threads+1 processor ids.
.PP
.B likwid-pin
supports different numberings for pinning. Per default physical numbering of
the cores is used.  This is the numbering also likwid-topology reports. But
also logical numbering inside the node or the sockets can be used.  If using
with a N (e.g. -c N:0-6) the cores are logical numbered over the whole node.
Physical cores come first. If a system e.g. has 8 cores with 16 SMT threads
with -c N:0-7 you get all physical cores.  If you specify -c N:0-15 you get all
physical cores and all SMT threads. With S you can specify logical numberings
inside sockets, again physical cores come first. You can mix different domains
with a @.  -c S0:0-3@S2:2-3 you pin thread 0-3 to logical cores 0-3 on socket 0
and threads 4-6 on logical cores 2-3 on socket 2.
.PP
For applications where first touch policy on numa systems cannot be employed
.B likwid-pin
can be used to turn on interleave memory placement. This can significantly
speed up the performance of memory bound multi threaded codes. All numa nodes
the user pinned threads to are used for interleaving.

.SH OPTIONS
.TP
.B \-\^v
prints version information to standard output, then exits.
.TP
.B \-\^h
prints a help message to standard output, then exits.
.TP
.B \-\^c " processor_list"
specify a numerical list of processors. The list may contain multiple 
items, separated by comma, and ranges. For example 0,3,9-11. You can also use
logical numberings, either within a node (N), a socket (S<id>) or a numa domain (M<id>).
likwid-pin also supports logical pinning within a cpuset with a L prefix.
.TP
.B \-\^t " OpenMP_type"
specify OpenMP implementation used. At the moment only intel is allowed as parameter.
.TP
.B \-\^s " skip_mask
Specify skip mask as HEX number. For each set bit the corresponding thread is skipped.
.TP
.B \-\^p
prints the available thread domains for logical pinning
.TP
.B \-\^i
set numa memory policy to interleave involving all numa nodes involved in pinning
.TP
.B \-\^q
silent execution without output


.SH EXAMPLE
.IP 1. 4
For  standard pthread application:
.TP
.B likwid-pin -c 0,2,4-6  ./myApp
.PP
The parent process is pinned to processor 0. Thread 0 to processor 2, thread
1 to processor 4, thread 2 to processor 5 and thread 3 to processor 6. If more threads
are created than specified in the processor list, these threads are pinned to processor 0
as fallback.
.IP 2. 4
For gcc OpenMP as many ids must be specified in processor list as there are threads: 
.TP
.B OMP_NUM_THREADS=4; likwid-pin -c 0,2,1,3  ./myApp
.IP 3. 4
For Intel icc OpenMP the flag 
.B \-\^t 
.I intel
must be set.
.TP
.B OMP_NUM_THREADS=4; likwid-pin -t intel -c S0:0,1@S1:0,1  ./myApp
.IP 4. 4
Full control over the pinning can be achieved by specifying a skip mask.
For example above case for Intel OpenMP can also be achieved with:
.TP
.B OMP_NUM_THREADS=4; likwid-pin -s 0x1 -c 0,2,1,3  ./myApp

.SH AUTHOR
Written by Jan Treibig <jan.treibig@gmail.com>.
.SH BUGS
Report Bugs on <http://code.google.com/p/likwid/issues/list>.
.SH "SEE ALSO"
taskset(1), likwid-perfctr(1), likwid-features(1), likwid-topology(1),


likwid/src/applications/likwid-pin.c
/*
 * ===========================================================================
 *
 *      Filename:  likwid-pin.c
 *
 *      Description:  An application to pin a program including threads
 *
 *      Version:  2.2
 *      Created:  14.06.2011
 *
 *      Author:  Jan Treibig (jt), jan.treibig@gmail.com
 *      Company:  RRZE Erlangen
 *      Project:  likwid
 *      Copyright:  Copyright (c) 2010, Jan Treibig
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License, v2, as
 *      published by the Free Software Foundation
 *     
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *     
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * ===========================================================================
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sched.h>
#include <errno.h>
#include <sys/types.h>
#include <unistd.h>
#include <ctype.h>

#include <error.h>
#include <types.h>
#include <bstrlib.h>
#include <cpuid.h>
#include <affinity.h>
#include <numa.h>
#include <strUtil.h>

#ifdef COLOR
#include <textcolor.h>
#endif


#define HELP_MSG \
printf("likwid-pin --  Version %d.%d \n\n",VERSION,RELEASE); \
printf("\n"); \
printf("Supported Options:\n"); \
printf("Example usage: likwid-pin -c 0,4-6 ./myApp\n"); \
printf("-h\t Help message\n"); \
printf("-v\t Version information\n"); \
printf("-i\t Set numa interleave policy with all involved numa nodes\n"); \
printf("-c\t comma separated processor ids\n"); \
printf("-s\t bitmask with threads to skip\n"); \
printf("-t\t Threading implementation type (e.g.intel)\n"); \
printf("-p\t Print available domains with mapping on physical ids\n"); \
printf("-q\t Silent without output\n\n"); \
printf("No special type necessary for gcc OpenMP\n");  \
printf("You can pin with the following numberings:\n");  \
printf("1. Physical numbering of OS.\n");  \
printf("2. Logical numbering inside node. e.g. -c N:0-3\n");  \
printf("3. Logical numbering inside socket. e.g. -c S0:0-3\n");  \
printf("4. Logical numbering inside last level cache group. e.g. -c C0:0-3\n");  \
printf("5. Logical numbering inside NUMA domain. e.g. -c M0:0-3\n\n");  \
printf("You can also mix domains separated by  @, e.g. -c S0:0-3@S1:0-3 \n\n")

#define VERSION_MSG \
printf("likwid-pin   %d.%d \n\n",VERSION,RELEASE)
    


void
pinPid(int cpuid, int silent)
{
    int status;
	cpu_set_t cpuset;

	CPU_ZERO(&cpuset);
	CPU_SET(cpuid, &cpuset);

	status = sched_setaffinity(0, sizeof(cpu_set_t), &cpuset);

	if (status == -1) 
	{
		printf("sched_setaffinity failed : %s \n",strerror(errno));
	}
	else 
    {
        if(!silent)
        {
#ifdef COLOR
            color_on(BRIGHT, COLOR);
#endif
            printf("[likwid-pin] Main PID -> core %d - OK\n",  cpuid);
#ifdef COLOR
            color_reset();
#endif
        }
    }
}


int main (int argc, char** argv)
{ 
    int i;
    int c;
	int skipMask = 0;
	int optInterleaved = 0;
	int optSilent = 0;
    bstring  typeString = bformat("NoType");
    bstring  pinString;
    bstring  skipString;
    bstring  argString;
    int numThreads=0;
    int threads[MAX_NUM_THREADS];
    threads[0] = 0;

    if (argc ==  1) { 
        HELP_MSG; 
        exit (EXIT_SUCCESS);    
    }

    cpuid_init();
    numa_init();
    affinity_init();

    while ((c = getopt (argc, argv, "+c:s:t:pqihv")) != -1)
    {
        switch (c)
        {
            case 'h':
                HELP_MSG;
                exit (EXIT_SUCCESS);    
            case 'v':
                VERSION_MSG;
                exit (EXIT_SUCCESS);    
            case 'i':
                optInterleaved = 1;
                break;
            case 'c':
                if (! (argString = bSecureInput(200,optarg)))
                {
                    ERROR_MSG(Failed to read argument string!);
                }

                numThreads = bstr_to_cpuset(threads, argString);

                if(!numThreads)
                {
                    ERROR_MSG(Failed to parse cpu list.);
                }

                bdestroy(argString);
                break;
            case 't':
                if (! (argString = bSecureInput(10,optarg)))
                {
                    ERROR_MSG(Failed to read argument string!);
                }

                typeString = bstrcpy(argString);
                bdestroy(argString);
                break;
            case 'p':
                affinity_printDomains();
                exit (EXIT_SUCCESS);    
                break;
            case 'q':
                optSilent = 1;
                setenv("LIKWID_SILENT","true", 1);
                break;
            case 's':
                if (! (argString = bSecureInput(10,optarg)))
                {
                    ERROR_MSG(Failed to read argument string!);
                }

                skipMask = strtoul((char*) argString->data,NULL,16);
                bdestroy(argString);
                break;
            default:
                HELP_MSG;
                exit(EXIT_FAILURE);
        }
    }

    if (!numThreads)
    {
        argString = bformat("N:0-%u", cpuid_topology.numHWThreads-1);
        numThreads = bstr_to_cpuset(threads, argString);
    }

	/* CPU List:
	 * pthread (default): pin main pid + all thread tids
	 *
	 * OpenMP: Pin OMP_NUM_THREADS
	 * intel openmp: pin main pid + all thread tids (skip thread 1)
	 * gcc openmp: pin main pid + all thread tids (one less)
	 */

    if (biseqcstr(typeString,"intel")) 
    {
		skipMask = 0x1;
        setenv("KMP_AFFINITY", "disabled", 1);
    }

    if (optInterleaved)
    {
        printf("Set mem_policy to interleaved\n");
        numa_setInterleaved(threads, numThreads);
    }

	if (numThreads > 1)
	{
		pinString = bformat("%d",threads[1]);

		for (i=2; i < numThreads;i++)
		{
			bformata(pinString,",%d",threads[i]);
		}

        skipString = bformat("%d",skipMask);

        setenv("LIKWID_PIN",(char*) pinString->data , 1);
        setenv("LIKWID_SKIP",(char*) skipString->data , 1);
        setenv("LD_PRELOAD",TOSTRING(LIBLIKWIDPIN), 1);
	}

	pinPid(threads[0], optSilent);
    fflush(stdout);

	argv +=  optind;
	execvp(argv[0], argv);
	perror("execvp");
	fprintf(stderr,"failed to execute %s\n", argv[0]);

    return EXIT_SUCCESS;
}

